In [93]:
import pandas as pd

Import all the files

In [94]:
visits = pd.read_csv('visits.csv',
                     parse_dates=[1])
cart = pd.read_csv('cart.csv',
                   parse_dates=[1])
                   
checkout = pd.read_csv('checkout.csv',
                       parse_dates=[1])
purchase = pd.read_csv('purchase.csv',
                       parse_dates=[1])

Step 1: Inspect the DataFrames using `print` and `head`

In [95]:
print(visits.head(5))
print(cart.head(5))
print(checkout.head(5))
print(purchase.head(5))

                                user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00
                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 

Step 2: Left merging visits and cart

In [96]:
visits_cart = pd.merge(visits, cart, how="left")
visits_cart.head()


user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3                 NaT  
4                 NaT

Step 3: How long is `visits_cart`?

In [97]:
visits_cart_length = len(visits_cart)
print(visits_cart_length)

2000


Step 4: How many timestamps are null for `cart_time`?

In [98]:
null_cart_time = visits_cart["cart_time"].isnull().sum()
print(null_cart_time)

1652


1652 people out of 2000 only visited the site.

Step 5: What percentage only visited?

In [99]:
only_visited_percentage = (null_cart_time/visits_cart_length)*100
print(only_visited_percentage)

82.6


More than 80% pf the people that visited did not proceed to put any T-shirt into their cart.

Step 6: What percentage placed a t-shirt in their cart but did not checkout?

In [100]:
cart_checkout = pd.merge(cart, checkout, how="left")
cart_checkout.head()

user_id           cart_time  \
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00   
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00   
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00   
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00   

        checkout_time  
0 2017-11-07 21:14:00  
1                 NaT  
2 2017-03-04 11:04:00  
3 2017-09-27 08:26:00  
4                 NaT

In [101]:
print(cart_checkout_length)

348


In [107]:
null_cart_checkout = cart_checkout["checkout_time"].isnull().sum()
print(null_cart_checkout)

122


In [109]:
null_cart_checkout_percentage = round((null_cart_checkout/cart_checkout_length)*100, 2)
print(null_cart_checkout_percentage)

35.06


Step 7: Merge it all together

In [115]:
all_data = visits.merge(cart, how="left").merge(cart, how="left").merge(checkout, how="left").merge(purchase, how="left")
all_data.head()


user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time       checkout_time       purchase_time  
0                 NaT                 NaT                 NaT  
1 2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  
2 2017-08-20 08:31:00                 NaT                 NaT  
3                 NaT                 NaT                 NaT  
4                 NaT                 NaT                 NaT

In [116]:
print(len(all_data))

2108


Step 8: % of users who got to checkout but did not purchase

In [92]:
checkout_users = all_data[all_data["checkout_time"].notnull()]
checkout_users.head()
print(len(checkout_users))

334


In [76]:
checkout_users_number = checkout_users["checkout_time"].notnull().sum()
print(checkout_users_number)

334


In [73]:
null_purchase_time = checkout_users["purchase_time"].isnull().sum()
print(null_purchase_time)

82


In [113]:
null_checkout_purchase_percentage = round((null_purchase_time/checkout_users_length)*100,2)
print(null_checkout_purchase_percentage)


24.55


Step 9: check each part of the funnel, let's print all 3 of them again

In [114]:
print(f"{only_visited_percentage} of users who visited did not put anything in their cart")
print(f"{null_cart_checkout_percentage} of users who put t-shirt in their cart did not proceed to check them out")
print(f"{null_checkout_purchase_percentage} of users who checked out did not proceed to make purchase")

82.6 of users who visited did not put anything in their cart
35.06 of users who put t-shirt in their cart did not proceed to check them out
24.55 of users who checked out did not proceed to make purchase


*The weakest part of the funnel is clearly getting a person who visited the site to add a tshirt to their cart. Once they've added a t-shirt to their cart it is fairly likely they end up purchasing it. A suggestion could be to make the add-to-cart button more prominent on the front page.*


Step 10: adding new column

In [121]:
checkout_users["time_spent"] = checkout_users["purchase_time"] - checkout_users["visit_time"]

Step 11: examine the results

In [122]:
checkout_users.head()

user_id          visit_time  \
1   0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
14  486480e2-98c3-4d51-8f4b-b1c07228ce84 2017-01-27 16:34:00   
31  1868e912-b38c-4ef0-8824-5665c42288e5 2017-03-17 11:48:00   
48  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:21:00   
49  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:21:00   

             cart_time       checkout_time       purchase_time      time_spent  
1  2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00 0 days 00:44:00  
14 2017-01-27 16:44:00 2017-01-27 17:10:00 2017-01-27 17:12:00 0 days 00:38:00  
31 2017-03-17 11:56:00 2017-03-17 11:58:00                 NaT             NaT  
48 2017-01-08 20:38:00 2017-01-08 20:52:00 2017-01-08 21:02:00 0 days 00:41:00  
49 2017-01-08 20:38:00 2017-01-08 20:52:00 2017-01-08 21:21:00 0 days 01:00:00

Step 12: average time to purchase

In [123]:
average_time = checkout_users["time_spent"].mean()
print(average_time)

0 days 00:43:12.380952380
